designs/0100-reduce_sum.md

```diff
@@ -0,0 +1,214 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+reduce_sum is effectively a parallel-for combined with a sum.
+
+For a set of input arguments, ```args0, args1, args2, ...``` and a scalar function ```f```, reduce_sum computes the scalar:
+
+```f(args0) + f(args1) + f(args2) + ...```
+
+In terms of probabilistic models, this is useful when there are N terms in a likelihood that combine multiplicatively. In this case, computing the log likelihood means computing the sum of the log of each of these terms. If a function ```f``` can be written that computes an individual log likelihood, then reduce_sum can be used to compute them all in parallel and accumulate the results.
+
+The actual meaning of N is flexible here. In a regression model, N might be the N different rows of the input dataframe, each which can be evaluated separately. In a hierarchical model, N might correspond to different groups -- perhaps there is some complex shared calculation for each group but these can all be done in parallel.
+
+reduce_sum is not useful, however, when there are dependencies between these likelihood terms. This can happen, for instance, if there were N terms in a Gaussian process likelihood. reduce_sum will not be useful for accelerating this.
+
+The signature we settled on for reduce_sum (which balances the different tradeoffs described in the Summary) is:
+
+```
+real reduce_sum(F func, T[] sliced_arg, int grainsize, T1 arg1, T2 arg2, ...)
+```
+
+where the function ```func``` has a signature like:
+
+```
+real func(int start, int end, T[] subset_sliced_arg, T1 arg1, T2 arg2, ...)
+```
+
+The array ```sliced_arg``` in the reduce_sum call represents the list of calculations to perform and sum.
+
+The user-provided function ```func``` is expect to compute the ```start``` through ```end``` terms of the overall sum, accumulate them, and return that value. The user function is only passed the subset ```sliced_arg[start:end]``` of sliced arg (as ```subset_sliced_arg```). ```start``` and ```end``` are passed so that ```func``` can index any of the ```argN``` appropriately. The trailing arguments ```argN``` are passed without modification to every call of ```func```.
+
+An overall call to ```reduce_sum``` can be replaced by either:
+
+```
+real sum = func(1, size(sliced_arg), sliced_arg, arg1, arg2, ...)
+```
+
+or (modulo differences due to rearrangements of summations) the code:
+
+```
+real sum = 0.0;
+for(i in 1:size(sliced_arg)) {
+  sum = sum + func(i, i, { sliced_arg[i] }, arg1, arg2, ...);
+}
+```
+
+The argument ```sliced_arg``` is called the sliced argument because func only receives the elements of ```sliced_arg``` for which it is responsible (in ```subset_sliced_arg```). All the other ```argN``` arguments are shared completely between every call to func.
```

Comment:
again, I think that the argument `sliced_arg` is actually `unsliced_arg` (or `items`) and `subset_sliced_arg` is `sliced_arg` (or `slice`)

Reply:
I think I'm explaining this in the wrong direction. I'll do a rewrite. It's hard to keep it all sorted.